pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,615 @@
+                    ArrayList<CompletableFuture<Void>> completableFutures = new ArrayList<>();
+                    CompletableFuture<Void> openManagedLedgerFuture = new CompletableFuture<>();
+                    AtomicBoolean hasInvalidIndex = new AtomicBoolean(false);
+                    AsyncCallbacks.OpenReadOnlyManagedLedgerCallback callback = new AsyncCallbacks
+                            .OpenReadOnlyManagedLedgerCallback() {
+                        @Override
+                        public void openReadOnlyManagedLedgerComplete(ReadOnlyManagedLedgerImpl readOnlyManagedLedger,
+                                                                      Object ctx) {
+                            finalPersistentSnapshotIndexes.getIndexList().forEach(index -> {
+                                CompletableFuture<Void> handleSegmentFuture = new CompletableFuture<>();
+                                completableFutures.add(handleSegmentFuture);
+                                readOnlyManagedLedger.asyncReadEntry(
+                                        new PositionImpl(index.getSegmentLedgerID(),
+                                                index.getSegmentEntryID()),
+                                        new AsyncCallbacks.ReadEntryCallback() {
+                                            @Override
+                                            public void readEntryComplete(Entry entry, Object ctx) {
+                                                handleSnapshotSegmentEntry(entry);
+                                                entry.release();
+                                                handleSegmentFuture.complete(null);
+                                            }
+
+                                            @Override
+                                            public void readEntryFailed(ManagedLedgerException exception, Object ctx) {
+                                                if (exception instanceof ManagedLedgerException
+                                                        .NonRecoverableLedgerException) {
+                                                    if (((ManagedLedgerImpl)topic.getManagedLedger())
+                                                            .ledgerExists(index.getAbortedMarkLedgerID())) {
+                                                        log.error("[{}] Failed to read snapshot segment [{}:{}]",
+                                                                topic.getName(), index.segmentLedgerID,
+                                                                index.segmentEntryID, exception);
+                                                        topic.close();
+                                                        handleSegmentFuture.completeExceptionally(exception);
+                                                    } else {
+                                                        indexes.remove(new PositionImpl(
```

Comment:
why not read first, read success put the persistent index into the memory index, read fail doesn't put